### PR TITLE
feat: limit the number of messages in prepare proposal

### PR DESCRIPTION
**Pull request by rach-id**
This PR limits the number of transactions in prepare proposal.

### CODE REVIEW

File: test/util/test_app.go
Comment (rach-id) [inline test/util/test_app.go:200]: @cmwaters I changed the max square size in here in this hacky way. If you think it's fine, I can make this PR ready for review.

File: app/validate_txs.go
Comment (evan-forbes) [inline app/validate_txs.go:64]: for posterity: iirc we're optimizing for simplicity here, and therefore simply counting the number of messages. in the future, we could use a multidimensional gas price model that would account for the savings of only verifying a single signature for multiple msgs

File: app/validate_txs.go
Comment (evan-forbes) [inline app/validate_txs.go:158]: very very optional

we might be able to refactor this and `msgTypes` into a single function that returns all of the different msg types in a map or smth

File: test/util/test_app.go
Comment (evan-forbes) [inline test/util/test_app.go:73]: > initiali~~s~~eTestApp

:zany_face: :innocent: 

zorro z

**Review by evan-forbes — COMMENTED** [review]
LGTM, waiting on others opinions before we merge

left a question about the ability to get past this check by changing the protobuf code that I'm unsure on

File: app/validate_txs.go
Comment (rach-id) [reply app/validate_txs.go:101]: if I understand correctly, the msg is taken after casting the sdk transaction. Do you mean we manually cast the transaction to a MsgSend and see? casts are often too expensive, that's why I just used the msg type URL

File: app/validate_txs.go
Comment (rach-id) [reply app/validate_txs.go:64]: https://github.com/celestiaorg/celestia-app/issues/3948

File: test/util/test_app.go
Comment (rach-id) [reply test/util/test_app.go:73]: British vs American 😆 . I guess I need to set my IDE to American English because I always need to fix the British differences

File: app/validate_txs.go
Comment (rach-id) [reply app/validate_txs.go:158]: we will need to sequentially go over the message types  in all cases, same complexity. Or you mean we make them part of a single function?

File: test/util/test_app.go
Comment (evan-forbes) [reply test/util/test_app.go:73]: yeah lol no changes needed!

File: app/validate_txs.go
Comment (evan-forbes) [reply app/validate_txs.go:158]: > Or you mean we make them part of a single function?

yeah, just this. very optional

Comment (rach-id) [reply app/validate_txs.go]: [54583ae](https://github.com/celestiaorg/celestia-app/pull/3942/commits/54583ae722bffa9a4c6756c17b3b00278ded1f3b)

File: app/validate_txs.go
Comment (rach-id) [inline app/validate_txs.go:56]: this is making the assumption that we can't mix PFBs and sdk messages in the same transaction. I guess that's the case currently if I am not missing anything

**Review by evan-forbes — APPROVED** [review]
LGTM, nice, thanks for pushing this through

**Review by rootulp — COMMENTED** [review]
I have a blocking question on whether we intend on applying this limit to v2 or just for v3. I think we just want it for v3 b/c that's why we're rushing to get it included in v3.

If that's the case, I think we need to add conditional logic so that the filtering only applies for app version >= 3. Perhaps worth adding a unit test that verifies v2 doesn't filter PFB or SDK messages greater than the constants.

BTW great existing unit tests in this PR 👍

File: app/test/prepare_proposal_test.go
Comment (rach-id) [reply app/test/prepare_proposal_test.go:286]: will revert this as apparently it's not affecting process proposal c59b82eff6d10ce6ababc65b358e4223d084ec11

File: app/test/prepare_proposal_test.go
Comment (rach-id) [reply app/test/prepare_proposal_test.go]: 230594ffa8b81726be35c424bece9e2af5bc2443

File: app/test/prepare_proposal_test.go
Comment (rach-id) [reply app/test/prepare_proposal_test.go]: c59b82eff6d10ce6ababc65b358e4223d084ec11

File: app/validate_txs.go
Comment (rach-id) [reply app/validate_txs.go]: 098f7cc97ccd049ec31ba444fe6f1c2c7a4495b2

File: app/validate_txs.go
Comment (rach-id) [reply app/validate_txs.go]: 098f7cc97ccd049ec31ba444fe6f1c2c7a4495b2

File: pkg/appconsts/global_consts.go
Comment (cmwaters) [inline pkg/appconsts/global_consts.go:78]: Can we include rationale for these numbers in the description. Ideally we also merge the benchmarks PR and link to those.

File: pkg/appconsts/global_consts.go
Comment (cmwaters) [inline pkg/appconsts/global_consts.go:74]: To make this clearer, can this be just `NonPFBTransactionCap` since PFBs are SDK msgs

File: pkg/appconsts/global_consts.go
Comment (cmwaters) [reply pkg/appconsts/global_consts.go:78]: I also feel like these numbers are overly conservative based on looking at the results. I had expected values in the low thousands.

File: test/util/test_app.go
Comment (cmwaters) [reply test/util/test_app.go:200]: There is a `genesis` package which looks like could do this functionality for you

File: test/util/test_app.go
Comment (cmwaters) [inline test/util/test_app.go:210]: I know this is somewhat legacy code but why do we return the application

File: test/util/test_app.go
Comment (cmwaters) [inline test/util/test_app.go:79]: Do you know if we normally call `Commit` before executing the first block i.e. `BeginBlock`

**Review by cmwaters — COMMENTED** [review]
Some of this testing infrastructure is getting a bit unwieldy and deserves a clean up.

The setup and running of a test app should have a canonical constructor and methods.

File: pkg/appconsts/global_consts.go
Comment (rach-id) [reply pkg/appconsts/global_consts.go:78]: For the rationale, it will be included after merging this PR and I will add comments, results and bencharmks etc.

For the numbers, the initial tests on my computer. But then we re-run the experiments on recommended validator config: 4cpu 16gb ram and we got lower numbers. So we reduced those to reflect a 0.25sec execution on that low config validator

File: test/util/test_app.go
Comment (rach-id) [reply test/util/test_app.go:79]: If we dont, prepare proposal fails, if I understand correctly

File: test/util/test_app.go
Comment (rach-id) [reply test/util/test_app.go:200]: Can you link me please to which part can do this for me?

File: app/test/prepare_proposal_test.go
Comment (rootulp) [inline app/test/prepare_proposal_test.go:270]: [optional][not blocking] currently this test only create transactions with a single message. It may be worth adding a test case with transactions that contain multiple messages to verify that the number of messages is the limiting constraint, not the number of transactions. After we do that, we could also:

```suggestion
func TestPrepareProposalCappingNumberOfMessages(t *testing.T) {
```

File: app/validate_txs.go
Comment (rootulp) [inline app/validate_txs.go:64]: [optional] rename for consistency `sdkTransactionsCount` => `nonPFBTransactionCount`

File: pkg/appconsts/global_consts.go
Comment (rootulp) [inline pkg/appconsts/global_consts.go:78]: [question][potentially blocking] I have a strong suspicion that we will need to modify these constraints in a future release. Should we make them versioned constants now to make it easier to modify them in future releases?

**Review by rootulp — APPROVED** [review]
LGTM with:
1. Question about converting the global constants into versioned constants
2. One extra test case for transactions with multiple messages

[optional] we may consider renaming the PR title to something more accurate based on Callum's comment since all messages are limited (not just msg send). One proposal:

"feat: limit the number of messages in prepare proposal"

File: pkg/appconsts/global_consts.go
Comment (rach-id) [reply pkg/appconsts/global_consts.go]: [5a59b88](https://github.com/celestiaorg/celestia-app/pull/3942/commits/5a59b882b2282460a75d539f3f3675fb8cc597be)

File: app/validate_txs.go
Comment (rach-id) [reply app/validate_txs.go]: 01e4cb4cf6c75a4567a9c88c609e32c0bcebd6c5

**Review by rootulp — APPROVED** [review]

**Review by ninabarbakadze — APPROVED** [review]
LGTM 🚢

**Comment by rach-id** [comment]
> One extra test case for transactions with multiple messages

c5071a3a3aaa8eea231492f1398123492b195508

File: app/validate_txs.go
Comment (evan-forbes) [reply app/validate_txs.go]: we're doing checks elsewhere when we decode the tx, so it should be okay. This is just something that always creeps me out as this string can be forged

Comment (rach-id) [reply app/validate_txs.go]: I see. If it makes sense, we can add a validation rule that the string should match the types at some point

File: test/util/test_app.go
Comment (evan-forbes) [reply test/util/test_app.go:200]: if we just want to change the square size, we can use https://github.com/celestiaorg/celestia-app/blob/3ea0ca5bd589d7212dbd429eaca8365ff2f35dc9/test/util/genesis/modifier.go#L24-L32

File: app/test/prepare_proposal_test.go
Comment (evan-forbes) [reply app/test/prepare_proposal_test.go:309]: at one point in time, we were able to set nonce etc using the application itself and didn't need an rpc, but I'm not sure that we have that infra any longer

**Review by evan-forbes — APPROVED** [review]
I'm good with merging this now and handling some testApp tech debt later this week

File: test/util/test_app.go
Comment (rach-id) [reply test/util/test_app.go:200]: https://github.com/celestiaorg/celestia-app/issues/3971

**Comment by rach-id** [comment]
https://github.com/celestiaorg/celestia-app/issues/3972

**Review by ninabarbakadze — APPROVED** [review]

File: pkg/appconsts/global_consts.go
Comment (evan-forbes) [reply pkg/appconsts/global_consts.go:78]: versioning is a good idea, as we could increase the limit upon upgrading. 

since this is in prepare proposal and its not consensus breaking, this is ofc not required. the example listed here for versioning could be applied in that upgrade

opening a FLUP so we can :ship:

Comment (evan-forbes) [reply pkg/appconsts/global_consts.go:78]: #3973

File: app/test/prepare_proposal_test.go
Comment (evan-forbes) [reply app/test/prepare_proposal_test.go:309]: resolving so we can merge but this was added as a flup

File: pkg/appconsts/global_consts.go
Comment (evan-forbes) [reply pkg/appconsts/global_consts.go]: resolving so we merge. @rach-id do you mind creating a PR for the benchmark code if you haven't already? we could also include the results from the last run to compare results

File: test/util/test_app.go
Comment (evan-forbes) [reply test/util/test_app.go:210]: iirc solely due to copy paste forking this code from the sdk

ref #2535

Comment (evan-forbes) [reply test/util/test_app.go:210]: resolving so we can merge

**Pull request merged**
